File: src/services/predictionService.ts
import { LoanApplication, PredictionResult } from '@/src/types'
import { MODEL_CONFIG, APPROVAL_CATEGORIES } from '@/src/constants'

export class PredictionService {
  static async predictDefault(application: LoanApplication): Promise<PredictionResult> {
    try {
      // Validate input data
      this.validateApplication(application)
      
      // Calculate risk score using weighted factors
      const riskScore = this.calculateRiskScore(application)
      
      // Convert to probability
      const defaultProbability = Math.min(Math.max(riskScore, 0.01), 0.99)
      
      // Calculate suggested rate
      const suggestedRate = this.calculateSuggestedRate(defaultProbability)
      
      // Determine risk level and approval status
      const riskLevel = this.getRiskLevel(defaultProbability)
      const approvalStatus = this.getApprovalStatus(defaultProbability)
      
      // Calculate confidence based on data completeness
      const confidence = this.calculateConfidence(application)
      
      return {
        defaultProbability,
        suggestedRate,
        riskLevel,
        confidence,
        approvalStatus
      }
    } catch (error) {
      throw new Error(`Prediction failed: ${error instanceof Error ? error.message : 'Unknown error'}`)
    }
  }

  private static validateApplication(app: LoanApplication): void {
    const required = ['age', 'monthlyIncome', 'debtRatio', 'creditUtilization']
    const missing = required.filter(field => app[field as keyof LoanApplication] == null)
    
    if (missing.length > 0) {
      throw new Error(`Missing required fields: ${missing.join(', ')}`)
    }
    
    if (app.age < 18 || app.age > 100) {
      throw new Error('Age must be between 18 and 100')
    }
    
    if (app.monthlyIncome <= 0) {
      throw new Error('Monthly income must be positive')
    }
  }

  private static calculateRiskScore(app: LoanApplication): number {
    let score = 0
    
    // Late payment history (highest weight)
    score += (app.late90Days || 0) * 0.35
    score += (app.late60Days || 0) * 0.25
    score += (app.late30Days || 0) * 0.15
    
    // Credit utilization
    score += Math.min(app.creditUtilization, 1.5) * 0.20
    
    // Debt-to-income ratio
    score += Math.min(app.debtRatio, 2.0) * 0.15
    
    // Age factor (younger = higher risk)
    if (app.age < 25) score += 0.08
    else if (app.age < 35) score += 0.04
    
    // Income factor
    if (app.monthlyIncome < 2500) score += 0.10
    else if (app.monthlyIncome < 4000) score += 0.05
    
    // Credit lines (too few or too many = risk)
    const creditLines = app.openCreditLines || 5
    if (creditLines < 3 || creditLines > 20) score += 0.05
    
    // Dependents
    score += Math.min((app.dependents || 0) * 0.02, 0.08)
    
    return score
  }

  private static calculateSuggestedRate(probability: number): number {
    const baseRate = MODEL_CONFIG.BASE_RATE
    const riskPremium = probability * MODEL_CONFIG.MAX_RISK_PREMIUM
    return Math.round((baseRate + riskPremium) * 10) / 10
  }

  private static getRiskLevel(probability: number): PredictionResult['riskLevel'] {
    if (probability < 0.3) return 'Low Risk'
    if (probability < 0.6) return 'Medium Risk'
    return 'High Risk'
  }

  private static getApprovalStatus(probability: number): PredictionResult['approvalStatus'] {
    if (probability < 0.3) return 'Prime'
    if (probability < 0.6) return 'Standard'
    if (probability < 0.8) return 'High-Rate'
    return 'Decline'
  }

  private static calculateConfidence(app: LoanApplication): number {
    const fields = Object.values(app).filter(v => v != null && v !== '')
    const completeness = fields.length / 10 // Total expected fields
    
    // Base confidence on data completeness and consistency
    let confidence = completeness * 0.8
    
    // Boost confidence for critical fields
    if (app.late90Days != null && app.late30Days != null) confidence += 0.1
    if (app.monthlyIncome > 0 && app.debtRatio > 0) confidence += 0.1
    
    return Math.min(confidence, 1.0)
  }
}

File: scripts/prediction_service.py
"""
Prediction Service
Loads the trained model and provides predictions via HTTP API
"""

import pickle
import pandas as pd
import numpy as np
from pathlib import Path
from flask import Flask, request, jsonify
from flask_cors import CORS

app = Flask(__name__)
CORS(app)  # Allow requests from Next.js frontend

# Load model and scaler on startup
model_dir = Path('public/models')
model = None
scaler = None

def load_model():
    global model, scaler
    try:
        with open(model_dir / 'model.pkl', 'rb') as f:
            model = pickle.load(f)
        with open(model_dir / 'scaler.pkl', 'rb') as f:
            scaler = pickle.load(f)
        print("✅ Model and scaler loaded successfully")
        return True
    except Exception as e:
        print(f"❌ Failed to load model: {e}")
        return False

@app.route('/predict', methods=['POST'])
def predict():
    try:
        data = request.json
        
        # Extract and prepare features in the same order as training
        features = {
            'RevolvingUtilizationOfUnsecuredLines': float(data['creditUtilization']),
            'NumberOfTime30-59DaysPastDueNotWorse': int(data['late30Days']),
            'DebtRatio': float(data['debtRatio']),
            'MonthlyIncome': float(data['monthlyIncome']),
            'NumberOfOpenCreditLinesAndLoans': int(data['openCreditLines']),
            'NumberOfTimes90DaysLate': int(data['late90Days']),
            'NumberRealEstateLoansOrLines': int(data['realEstateLoans']),
            'NumberOfTime60-89DaysPastDueNotWorse': int(data['late60Days']),
            'NumberOfDependents': int(data['dependents']),
        }
        
        # Calculate engineered features (same as training)
        features['DebtRatio_Capped'] = min(features['DebtRatio'], np.percentile([features['DebtRatio']], 95))
        features['TotalCreditLines'] = features['NumberOfOpenCreditLinesAndLoans'] + features['NumberRealEstateLoansOrLines']
        features['UtilizationRate'] = min(features['RevolvingUtilizationOfUnsecuredLines'], 1.5)
        features['LatePaymentRisk'] = (
            features['NumberOfTime30-59DaysPastDueNotWorse'] * 1 +
            features['NumberOfTime60-89DaysPastDueNotWorse'] * 2 +
            features['NumberOfTimes90DaysLate'] * 3
        )
        
        # Create feature array in correct order
        feature_order = [
            'RevolvingUtilizationOfUnsecuredLines', 'NumberOfTime30-59DaysPastDueNotWorse',
            'DebtRatio', 'MonthlyIncome', 'NumberOfOpenCreditLinesAndLoans',
            'NumberOfTimes90DaysLate', 'NumberRealEstateLoansOrLines',
            'NumberOfTime60-89DaysPastDueNotWorse', 'NumberOfDependents',
            'DebtRatio_Capped', 'TotalCreditLines', 'UtilizationRate', 'LatePaymentRisk'
        ]
        
        X = np.array([[features[col] for col in feature_order]])
        
        # Scale features
        X_scaled = scaler.transform(X)
        
        # Make prediction
        probabilities = model.predict_proba(X_scaled)[0]
        probability = probabilities[1]  # Probability of default
        
        # Calculate confidence score (how certain the model is)
        confidence = max(probabilities) * 100  # Higher when model is more certain
        
        # Calculate optimal interest rate based on risk
        base_rate = 5.5  # Base prime rate
        risk_premium = probability * 15  # Risk-based premium (0-15%)
        suggested_rate = base_rate + risk_premium
        
        # Determine risk level and pricing strategy
        if probability >= 0.6:
            risk_level = "High Risk"
            if probability >= 0.8:
                recommendation = f"✗ Decline - Default risk too high ({probability*100:.1f}%)"
                pricing_decision = "DECLINE"
                suggested_rate = None
            else:
                recommendation = f"⚠ High-Risk Approval - Approve at {suggested_rate:.1f}% APR"
                pricing_decision = "APPROVE_HIGH_RATE"
        elif probability >= 0.3:
            risk_level = "Medium Risk"
            recommendation = f"⚠ Standard Approval - Approve at {suggested_rate:.1f}% APR"
            pricing_decision = "APPROVE_STANDARD"
        else:
            risk_level = "Low Risk"
            recommendation = f"✓ Prime Approval - Approve at {suggested_rate:.1f}% APR"
            pricing_decision = "APPROVE_PRIME"
        
        # Identify top risk factors
        top_factors = []
        if features['NumberOfTimes90DaysLate'] > 0:
            top_factors.append({"factor": "90+ Days Late Payments", "impact": "high"})
        if features['RevolvingUtilizationOfUnsecuredLines'] > 0.5:
            top_factors.append({"factor": "High Credit Utilization", "impact": "high"})
        if features['DebtRatio'] > 0.5:
            top_factors.append({"factor": "High Debt-to-Income Ratio", "impact": "high"})
        if features['MonthlyIncome'] < 3000:
            top_factors.append({"factor": "Low Monthly Income", "impact": "medium"})

        
        if not top_factors:
            top_factors.append({"factor": "Clean Payment History", "impact": "low"})
            top_factors.append({"factor": "Healthy Credit Utilization", "impact": "low"})
        
        return jsonify({
            "defaultProbability": float(probability),
            "confidence": float(confidence),
            "riskLevel": risk_level,
            "recommendation": recommendation,
            "suggestedRate": float(suggested_rate) if suggested_rate else None,
            "pricingDecision": pricing_decision,
            "topFactors": top_factors[:5]
        })
        
    except Exception as e:
        print(f"Prediction error: {e}")
        return jsonify({"error": "Failed to generate prediction"}), 500

@app.route('/health', methods=['GET'])
def health():
    return jsonify({"status": "healthy", "model_loaded": model is not None})

if __name__ == '__main__':
    if load_model():
        print("🚀 Starting prediction service on http://localhost:5000")
        app.run(host='0.0.0.0', port=5000, debug=True)
    else:
        print("❌ Cannot start service - model loading failed")